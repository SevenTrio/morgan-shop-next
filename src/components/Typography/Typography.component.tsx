import { forwardRef, HTMLAttributes } from 'react';

import { colorVariant } from 'types/common.types';
import { ISpacingProps } from 'types/spacing.types';
import {
  typographyAlign,
  typographyComponent,
  typographyVariant,
} from 'types/typography.types';

import {
  Headline1,
  Body1,
  Body2,
  Body3,
  ButtonText,
  Caption1,
  Caption2,
  Caption3,
} from './Typography.styles';

export interface ITypographyProps
  extends ISpacingProps,
    HTMLAttributes<HTMLParagraphElement> {
  /**
   * Set the text-align on the component.
   */
  align?: typographyAlign;
  /**
   * The color of the component.
   */
  color?: colorVariant;
  /**
   * The component used for the root node.
   */
  component?: typographyComponent;
  /**
   * If true, the text will not wrap, but instead will truncate with a text overflow ellipsis.
   */
  noWrap?: boolean;
  /**
   * The variant to use.
   */
  variant?: typographyVariant;
}

export const Typography = forwardRef<HTMLParagraphElement, ITypographyProps>(
  (
    {
      align = 'inherit',
      color = 'textPrimary',
      variant = 'body1',
      component,
      noWrap = false,
      children,
      ...restProps
    },
    ref,
  ) => {
    switch (variant) {
      case `h1`:
        return (
          <Headline1
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Headline1>
        );

      case `body1`:
        return (
          <Body1
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Body1>
        );

      case `body2`:
        return (
          <Body2
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Body2>
        );

      case `body3`:
        return (
          <Body3
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Body3>
        );

      case `buttonText`:
        return (
          <ButtonText
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </ButtonText>
        );

      case `caption1`:
        return (
          <Caption1
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Caption1>
        );

      case `caption2`:
        return (
          <Caption2
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Caption2>
        );

      case `caption3`:
        return (
          <Caption3
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Caption3>
        );

      default:
        return (
          <Body1
            ref={ref}
            as={component}
            align={align}
            color={color}
            noWrap={noWrap}
            {...restProps}
          >
            {children}
          </Body1>
        );
    }
  },
);
