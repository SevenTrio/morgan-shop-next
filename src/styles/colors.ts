const colors = {
  black: `#121212`,
  darkGray: `#6D6D6D`,
  gray: `#DEDEDE`,
  lightGray: `#F0F0F0`,
  white: `#FAFAFA`,
  red: `#D92222`,
};

export default colors;
