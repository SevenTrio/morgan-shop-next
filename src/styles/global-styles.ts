import { createGlobalStyle } from 'styled-components';

export const GlobalStyle = createGlobalStyle`
  html {
    box-sizing: border-box;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  body {
    font-size: 16px;
    font-family: 'Poppins', sans-serif;
  }

  img {
    display: block;
    max-width: 100%;
    overflow: hidden;
  }

  p,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 0;
  }
`;
