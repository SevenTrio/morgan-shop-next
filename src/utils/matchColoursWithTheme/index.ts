import { matchColoursWithTheme } from './matchColoursWithTheme';

export default matchColoursWithTheme;
