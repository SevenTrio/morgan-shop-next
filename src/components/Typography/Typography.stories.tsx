import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { Typography } from './Typography.component';

export default {
  title: `Components/Typography`,
  component: Typography,
} as ComponentMeta<typeof Typography>;

const Template: ComponentStory<typeof Typography> = (args) => (
  <Typography {...args} />
);

export const Headline1 = Template.bind({});
Headline1.args = {
  children: `Headline1. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `h1`,
};

export const Body1 = Template.bind({});
Body1.args = {
  children: `Body1. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `body1`,
};

export const Body2 = Template.bind({});
Body2.args = {
  children: `Body2. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `body2`,
};

export const Body3 = Template.bind({});
Body3.args = {
  children: `Body3. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `body3`,
};

export const ButtonText = Template.bind({});
ButtonText.args = {
  children: `ButtonText. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `buttonText`,
};

export const Caption1 = Template.bind({});
Caption1.args = {
  children: `Caption1. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `caption1`,
};

export const Caption2 = Template.bind({});
Caption2.args = {
  children: `Caption2. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `caption2`,
};

export const Caption3 = Template.bind({});
Caption3.args = {
  children: `Caption3. Lorem ipsum dolor sit amet, consectetur adipisicing elit.`,
  variant: `caption3`,
};
