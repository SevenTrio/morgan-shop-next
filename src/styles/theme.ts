import { DefaultTheme } from 'styled-components';
import { lighten, darken } from 'polished';

import spacing from 'utils/spacing';
import colors from 'styles/colors';

export const theme: DefaultTheme = {
  breakpoints: {
    xs: 0,
    sm: 600,
    md: 960,
    lg: 1280,
    xl: 1920,
  },
  palette: {
    common: {
      black: colors.black,
      darkGray: colors.darkGray,
      gray: colors.gray,
      lightGray: colors.lightGray,
      white: colors.white,
      red: colors.red,
    },
    primary: {
      light: `#424242`,
      main: colors.black,
      dark: darken(0.08, colors.black),
      contrastText: colors.white,
    },
    secondary: {
      light: lighten(0.08, colors.white),
      main: colors.white,
      dark: darken(0.08, colors.white),
      contrastText: colors.black,
    },
    error: {
      light: lighten(0.08, colors.red),
      main: colors.red,
      dark: darken(0.08, colors.red),
      contrastText: colors.white,
    },
    text: {
      primary: colors.black,
      secondary: colors.darkGray,
      disabled: colors.darkGray,
      hint: colors.darkGray,
    },
    divider: colors.gray,
    background: {
      default: colors.white,
      productCard: colors.lightGray,
      menu: colors.lightGray,
    },
  },
  shadows: [`1px 1px 10px rgba(0, 0, 0, 0.08)`],
  spacing,
  zIndexes: {
    basket: 100,
  },
};
