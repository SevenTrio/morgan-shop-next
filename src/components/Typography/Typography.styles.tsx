import styled, { css } from 'styled-components';

import { colorVariant } from 'types/common.types';
import { ISpacingProps } from 'types/spacing.types';
import { typographyAlign } from 'types/typography.types';

import spacingStyles from 'utils/spacingStyles';
import matchColoursWithTheme from 'utils/matchColoursWithTheme';

interface ITypographyCommonStylesProps extends ISpacingProps {
  align: typographyAlign;
  color: colorVariant;
  noWrap: boolean;
}

const typographyCommonStyles = css<ITypographyCommonStylesProps>`
  ${spacingStyles};

  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: ${({ theme, color }) => matchColoursWithTheme(theme, color)};

  ${({ align }) =>
    align &&
    css`
      text-align: ${align};
    `}

  ${({ noWrap }) =>
    noWrap &&
    css`
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    `}
`;

export const Headline1 = styled.h1`
  ${typographyCommonStyles};
  font-size: 1.125rem;
`;

export const Body1 = styled.p`
  ${typographyCommonStyles};
  font-weight: 500;
`;

export const Body2 = styled.p`
  ${typographyCommonStyles};
  font-size: 0.875rem;
`;

export const Body3 = styled.p`
  ${typographyCommonStyles};
  font-size: 0.875rem;
  font-weight: 500;
`;

export const ButtonText = styled.p`
  ${typographyCommonStyles};
  letter-spacing: 0.02em;
`;

export const Caption1 = styled.span`
  ${typographyCommonStyles};
  font-size: 0.75rem;
  line-height: 1.6666666666;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.sm}px) {
    font-size: 0.625rem;
    line-height: 1.6;
  }
`;

export const Caption2 = styled.span`
  ${typographyCommonStyles};
  font-size: 0.75rem;
  font-weight: 500;
  line-height: 1.6666666666;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.sm}px) {
    font-size: 0.625rem;
    line-height: 1.6;
  }
`;

export const Caption3 = styled.span`
  ${typographyCommonStyles};
  font-size: 0.75rem;
  text-decoration-line: underline;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.sm}px) {
    font-size: 0.625rem;
  }
`;
