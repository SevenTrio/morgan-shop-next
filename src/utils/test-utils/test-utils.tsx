import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { Normalize } from 'styled-normalize';

import { GlobalStyle } from 'styles/global-styles';
import { theme } from 'styles/theme';

const AllTheProviders: FC<{ children: React.ReactNode }> = ({ children }) => (
  <ThemeProvider theme={theme}>
    <Normalize />
    <GlobalStyle />
    {children}
  </ThemeProvider>
);

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
