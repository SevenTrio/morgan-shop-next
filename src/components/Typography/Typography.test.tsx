import { render } from 'utils/test-utils';
import { theme } from 'styles/theme';

import Typography from 'components/Typography';

describe(`Typography component`, () => {
  describe(`renders correctly`, () => {
    test(`without props`, () => {
      const { container } = render(<Typography />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
    });

    test(`with children prop`, () => {
      const typographyText = `Lorem ipsum dolor.`;
      const { container } = render(<Typography>{typographyText}</Typography>);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree?.textContent).toBe(typographyText);
    });
  });

  describe(`applies correct styles`, () => {
    test(`by default`, () => {
      const { container } = render(<Typography />);
      const tree = container.firstChild;

      expect(tree).toHaveStyleRule(`padding`, `0`);
      expect(tree).toHaveStyleRule(`font-size`, `1rem`);
      expect(tree).toHaveStyleRule(`font-weight`, `500`);
      expect(tree).toHaveStyleRule(`line-height`, `1.5`);
      expect(tree).toHaveStyleRule(`color`, theme.palette.text.primary);
      expect(tree).not.toHaveStyleRule(`text-align`);
      expect(tree).not.toHaveStyleRule(`overflow`);
      expect(tree).not.toHaveStyleRule(`white-space`);
      expect(tree).not.toHaveStyleRule(`text-overflow`);
    });

    test(`with variant prop`, () => {
      const { container } = render(<Typography variant="body3" />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`font-size`, `0.875rem`);
      expect(tree).toHaveStyleRule(`font-weight`, `500`);
    });

    test(`with color prop`, () => {
      const { container } = render(<Typography color="textSecondary" />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`color`, theme.palette.text.secondary);
    });

    test(`with align prop`, () => {
      const { container } = render(<Typography align="center" />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`text-align`, `center`);
    });

    test(`to wrapper component`, () => {
      const { container } = render(
        <Typography m={2} style={{ display: 'none' }} className="Test" />,
      );
      const tree = container.firstChild as HTMLElement;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`margin`, `8px`);
      expect(tree?.style.display).toBe('none');
      expect(tree?.className.split(' ').includes('Test')).toBe(true);
    });
  });

  describe(`renders correct tag`, () => {
    test(`with component prop`, () => {
      const { container } = render(<Typography component="h1" />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree?.nodeName).toBe(`H1`);
    });

    test(`with variant prop`, () => {
      const { container } = render(<Typography variant="caption1" />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree?.nodeName).toBe(`SPAN`);
    });

    test(`with both component and variant props`, () => {
      const { container } = render(
        <Typography component="h1" variant="caption1" />,
      );
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree?.nodeName).toBe(`H1`);
    });
  });
});
