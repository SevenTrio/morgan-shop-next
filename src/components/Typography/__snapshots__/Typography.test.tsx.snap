// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography component applies correct styles to wrapper component 1`] = `
.c0 {
  margin: 8px;
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  font-weight: 500;
}

<p
  class="c0 Test"
  color="textPrimary"
  style="display: none;"
/>
`;

exports[`Typography component applies correct styles with align prop 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: center;
  font-weight: 500;
}

<p
  class="c0"
  color="textPrimary"
/>
`;

exports[`Typography component applies correct styles with color prop 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #6D6D6D;
  text-align: inherit;
  font-weight: 500;
}

<p
  class="c0"
  color="textSecondary"
/>
`;

exports[`Typography component applies correct styles with variant prop 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  font-size: 0.875rem;
  font-weight: 500;
}

<p
  class="c0"
  color="textPrimary"
/>
`;

exports[`Typography component renders correct tag with both component and variant props 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  font-size: 0.75rem;
  line-height: 1.6666666666;
}

@media screen and (max-width:600px) {
  .c0 {
    font-size: 0.625rem;
    line-height: 1.6;
  }
}

<h1
  class="c0"
  color="textPrimary"
/>
`;

exports[`Typography component renders correct tag with component prop 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  font-weight: 500;
}

<h1
  class="c0"
  color="textPrimary"
/>
`;

exports[`Typography component renders correct tag with variant prop 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  font-size: 0.75rem;
  line-height: 1.6666666666;
}

@media screen and (max-width:600px) {
  .c0 {
    font-size: 0.625rem;
    line-height: 1.6;
  }
}

<span
  class="c0"
  color="textPrimary"
/>
`;

exports[`Typography component renders correctly with children prop 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  font-weight: 500;
}

<p
  class="c0"
  color="textPrimary"
>
  Lorem ipsum dolor.
</p>
`;

exports[`Typography component renders correctly without props 1`] = `
.c0 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  font-weight: 500;
}

<p
  class="c0"
  color="textPrimary"
/>
`;
